### ISSUE TITLE
gcp.datastream.ConnectionProfile not saving bigqueryProfile in State

### ISSUE THREAD

**Issue by breno-bicycle** [pulumi/pulumi-gcp#1874]
### What happened?

We need to upgrade our @pulumi/gcp package to be able to use the `args.settings.ipConfiguration.sslMode` property in `new sql.DatabaseInstance`.

After upgrading the @pulumi/gcp package version from 6.65.0 to latest (7.15.0) and running `pulumi up`, it updated a lot of resources due to the version diff successfully, but the `datastream.ConnectionProfile` failed. Looking at the `--diff` we can see that it is trying to add the bigqueryProfile property, that should be already there, and failing. I deleted the resource then recreated successfully. After running again, it tried again to add the `bigqueryProfile`, and fail.

### Example

Code
```
const bigQueryReplicationConnectionProfile = new gcp.datastream.ConnectionProfile(
    `${dbName}_bigquery-replication-connection-profile`,
    {
        displayName: `${dbName}_bigquery-replication-cp`,
        location: 'us-central1',
        connectionProfileId: `${dbName}_bigquery-replication-cp`,
        bigqueryProfile: {},
    },
    {
        dependsOn: [databaseInstance, replicationSlot, postgresReplicationConnectionProfile],
        parent: this,
    }
);
```
Diff
![Screen Shot 2024-03-25 at 19 06 09](https://github.com/pulumi/pulumi-gcp/assets/95237104/c86086bb-64e6-40b5-9e5d-a5cac07c076c)
Error
![Screen Shot 2024-03-25 at 19 02 21](https://github.com/pulumi/pulumi-gcp/assets/95237104/6fa54643-9936-4d04-b861-dc16486d99e6)

### Output of `pulumi about`

```
CLI          
Version      3.75.0
Go Version   go1.20.6
Go Compiler  gc

Plugins
NAME        VERSION
command     0.9.2
gcp         7.15.0
nodejs      unknown
postgresql  3.11.0
random      4.16.0

Host     
OS       darwin
Version  12.6.7
Arch     arm64

This project is written in nodejs: executable='/opt/homebrew/opt/node@18/bin/node' version='v18.17.0'
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by VenelinMartinov** [comment]
Sorry you are facing this issue and thanks for reporting! We'll try to get to this as soon as we can.

Notes for development here: the upstream schema is somewhat odd, so we might not be handling it correcty:

https://github.com/hashicorp/terraform-provider-google-beta/blob/fdd9e2a24142f05985d8d102ab2444851f332e4e/google-beta/services/datastream/resource_datastream_connection_profile.go#L74

```golang
"bigquery_profile": {
		Type:        schema.TypeList,
		Optional:    true,
		Description: `BigQuery warehouse profile.`,
		MaxItems:    1,
		Elem: &schema.Resource{
			Schema: map[string]*schema.Schema{},
		},
		ExactlyOneOf: []string{"oracle_profile", "gcs_profile", "mysql_profile", "bigquery_profile", "postgresql_profile"},
	}
```

This could also be related to https://github.com/pulumi/pulumi-terraform-bridge/issues/1767

**Comment by breno-bicycle** [comment]
Just as context, the version 6.65.0 creates and updates the bigquery ConnectionProfile with no issues. My issue there is the sslMode that I need to use `ENCRYPTED_ONLY` and there is no way to have same result in that version

**Comment by mikhailshilkov** [comment]
@VenelinMartinov I'm marking this as a P1 and regression since it worked in v6 but doesn't in v7.

**Comment by VenelinMartinov** [comment]
Full repro:

`index.ts`:

```typescript
import * as pulumi from "@pulumi/pulumi";
import * as gcp from "@pulumi/gcp";

new gcp.datastream.ConnectionProfile(
    `db-replication-connection-profile`,
    {
        displayName: `$db_bigquery-replication-cp`,
        location: 'us-central1',
        connectionProfileId: `db_bigquery-replication-cp`,
        bigqueryProfile: {},
    },
);
```

`package.json`:

```json
{
    "name": "gcp_conn_profile",
    "main": "index.ts",
    "devDependencies": {
        "@types/node": "^18"
    },
    "dependencies": {
        "@pulumi/pulumi": "^3.0.0",
        "@pulumi/gcp": "^6.0.0"
    }
}
```

`repro.sh`:

```bash
#!/bin/bash
set -euxo pipefail
pulumi up --yes
sed -i -e 's/"^6.0.0"/"^7.0.0"/g' package.json
pulumi up --yes
```

Yields the right error:

```
Diagnostics:
  pulumi:pulumi:Stack (gcp_conn_profile-dev):
    error: update failed

  gcp:datastream:ConnectionProfile (db-replication-connection-profile):
    error: 1 error occurred:
    	* updating urn:pulumi:dev::gcp_conn_profile::gcp:datastream/connectionProfile:ConnectionProfile::db-replication-connection-profile: 1 error occurred:
    	* Error waiting for Updating ConnectionProfile: {"@type":"type.googleapis.com/google.rpc.ErrorInfo","domain":"datastream.googleapis.com","metadata":{"message":"An unknown error occurred. Please try again. If the error persists, contact Google support.","originalMessage":"","time":"2024-04-01T12:09:08Z","uuid":"a4a5731d-f020-11ee-932f-5a67a52ffce7"},"reason":"UNKNOWN"}
    {"code":"VALIDATE_CONNECTIVITY","description":"Validates that Datastream can connect to the source database.","message":[{"code":"INTERNAL_ERROR","level":"ERROR","metadata":{"original_error":""}}],"state":"FAILED"}
```

Note that the same does not happen if we provision the resource on 7.x the first time, so looks like an upgrade issue.

**Comment by VenelinMartinov** [comment]
There might be two different errors here:

1. We produce a diff for the `bigqueryProfile` parameter in the first place. This one has no reason to happen as far as I can see:
```json
{
    "method": "/pulumirpc.ResourceProvider/Diff",
    "request": {
        "id": "projects/pulumi-development/locations/us-central1/connectionProfiles/db_bigquery-replication-cp-FZz3PEgE6Hx3iUgi",
        "urn": "urn:pulumi:dev::connection-profile::gcp:datastream/connectionProfile:ConnectionProfile::dbReplicationConnectionProfile",
        "olds": {
            "__meta": "{\"e2bfb730-ecaa-11e6-8f88-34363bc7c4c0\":{\"create\":1200000000000,\"delete\":1200000000000,\"update\":1200000000000}}",
            "bigqueryProfile": null,
            "connectionProfileId": "db_bigquery-replication-cp-FZz3PEgE6Hx3iUgi",
            "displayName": "db_bigquery-replication-cp",
            "forwardSshConnectivity": null,
            "gcsProfile": null,
            "id": "projects/pulumi-development/locations/us-central1/connectionProfiles/db_bigquery-replication-cp-FZz3PEgE6Hx3iUgi",
            "labels": {},
            "location": "us-central1",
            "mysqlProfile": null,
            "name": "projects/pulumi-development/locations/us-central1/connectionProfiles/db_bigquery-replication-cp-FZz3PEgE6Hx3iUgi",
            "oracleProfile": null,
            "postgresqlProfile": null,
            "privateConnectivity": null,
            "project": "pulumi-development"
        },
        "news": {
            "__defaults": [],
            "bigqueryProfile": {
                "__defaults": []
            },
            "connectionProfileId": "db_bigquery-replication-cp-FZz3PEgE6Hx3iUgi",
            "displayName": "db_bigquery-replication-cp",
            "location": "us-central1"
        },
        "oldInputs": {
            "__defaults": [],
            "bigqueryProfile": {
                "__defaults": []
            },
            "connectionProfileId": "db_bigquery-replication-cp-FZz3PEgE6Hx3iUgi",
            "displayName": "db_bigquery-replication-cp",
            "location": "us-central1"
        }
    },
    "response": {
        "stables": [
            "connectionProfileId",
            "location",
            "project"
        ],
        "changes": "DIFF_SOME",
        "diffs": [
            "bigqueryProfile"
        ],
        "detailedDiff": {
            "bigqueryProfile": {
                "kind": "UPDATE"
            }
        },
        "hasDetailedDiff": true
    },
    "metadata": {
        "kind": "resource",
        "mode": "client",
        "name": "gcp"
    }
}
```

2. The GCP provider does not handle whatever input we send for the update well.

```json
{
    "method": "/pulumirpc.ResourceProvider/Update",
    "request": {
        "id": "projects/pulumi-development/locations/us-central1/connectionProfiles/db_bigquery-replication-cp-FZz3PEgE6Hx3iUgi",
        "urn": "urn:pulumi:dev::connection-profile::gcp:datastream/connectionProfile:ConnectionProfile::dbReplicationConnectionProfile",
        "olds": {
            "__meta": "{\"e2bfb730-ecaa-11e6-8f88-34363bc7c4c0\":{\"create\":1200000000000,\"delete\":1200000000000,\"update\":1200000000000}}",
            "bigqueryProfile": null,
            "connectionProfileId": "db_bigquery-replication-cp-FZz3PEgE6Hx3iUgi",
            "displayName": "db_bigquery-replication-cp",
            "forwardSshConnectivity": null,
            "gcsProfile": null,
            "id": "projects/pulumi-development/locations/us-central1/connectionProfiles/db_bigquery-replication-cp-FZz3PEgE6Hx3iUgi",
            "labels": {},
            "location": "us-central1",
            "mysqlProfile": null,
            "name": "projects/pulumi-development/locations/us-central1/connectionProfiles/db_bigquery-replication-cp-FZz3PEgE6Hx3iUgi",
            "oracleProfile": null,
            "postgresqlProfile": null,
            "privateConnectivity": null,
            "project": "pulumi-development"
        },
        "news": {
            "__defaults": [],
            "bigqueryProfile": {
                "__defaults": []
            },
            "connectionProfileId": "db_bigquery-replication-cp-FZz3PEgE6Hx3iUgi",
            "displayName": "db_bigquery-replication-cp",
            "location": "us-central1"
        },
        "oldInputs": {
            "__defaults": [],
            "bigqueryProfile": {
                "__defaults": []
            },
            "connectionProfileId": "db_bigquery-replication-cp-FZz3PEgE6Hx3iUgi",
            "displayName": "db_bigquery-replication-cp",
            "location": "us-central1"
        }
    },
    "errors": [
        "rpc error: code = Unknown desc = updating urn:pulumi:dev::connection-profile::gcp:datastream/connectionProfile:ConnectionProfile::dbReplicationConnectionProfile: 1 error occurred:\n\t* Error waiting for Updating ConnectionProfile: {\"@type\":\"type.googleapis.com/google.rpc.ErrorInfo\",\"domain\":\"datastream.googleapis.com\",\"metadata\":{\"message\":\"An unknown error occurred. Please try again. If the error persists, contact Google support.\",\"originalMessage\":\"\",\"time\":\"2024-04-01T13:29:42Z\",\"uuid\":\"e5a83362-f02b-11ee-932f-5a67a52ffce7\"},\"reason\":\"UNKNOWN\"}\n{\"code\":\"VALIDATE_CONNECTIVITY\",\"description\":\"Validates that Datastream can connect to the source database.\",\"message\":[{\"code\":\"INTERNAL_ERROR\",\"level\":\"ERROR\",\"metadata\":{\"original_error\":\"\"}}],\"state\":\"FAILED\"}\n\n\n\n"
    ],
    "metadata": {
        "kind": "resource",
        "mode": "client",
        "name": "gcp"
    }
}
```

Not sure why we are running into this but it seems like it might an issue upstream and we wouldn't be hitting if 1. didn't happen.

**Comment by VenelinMartinov** [comment]
Not an upgrade issue after all - we just always produce a diff for the `bigqueryProfile` parameter and updating the resource fails after. Reproes with just v7.

The same thing does not happen with 6.67.0.

**Comment by VenelinMartinov** [comment]
Not fixed by enrolling the resource into PlanResourceChange - I hacked it up to get around https://github.com/pulumi/pulumi-terraform-bridge/issues/1822 by adding a `return nil, nil` in https://github.com/pulumi/pulumi-terraform-bridge/blob/f6013b69e741bc73632fd42ba2b4630a1f2cc3ff/pkg/tfshim/sdk-v2/provider2.go#L284

**Comment by VenelinMartinov** [comment]
Tried this in TF, does not reproduce:

```terraform
provider "google" {
  region = "us-central1"
}

terraform {
  required_providers {
    google = {
      source  = "hashicorp/google"
      version = "4.84.0"
    }
  }

}

resource "google_datastream_connection_profile" "default" {
    display_name          = "Connection profile"
    location              = "us-central1"
    connection_profile_id = "my-profile"
    bigquery_profile {}
}
```

Certainly seems like a bridge issue. Tried to repro with a bridge cross-test but failed so far https://github.com/pulumi/pulumi-terraform-bridge/pull/1823

**Comment by VenelinMartinov** [comment]
Fairly sure we are mishandling the `bigquery_profile` when we create the resource.

Dumped the `olds` when diffing and got:

```
olds=resource.PropertyMap{
           "__meta": {
               V:  "{\"e2bfb730-ecaa-11e6-8f88-34363bc7c4c0\":{\"create\":1200000000000,\"delete\":1200000000000,\"update\":1200000000000}}",
           },
           "connectionProfileId": {
               V:  "db_bigquery-replication-cp-FZz3PEgE6Hx3iUgi",
           },
           "displayName": {
               V:  "db_bigquery-replication-cp",
           },
           "id": {
               V:  "projects/pulumi-development/locations/us-central1/connectionProfiles/db_bigquery-replication-cp-FZz3PEgE6Hx3iUgi",
           },
           "labels": {
               V:  resource.PropertyMap{
               },
           },
           "location": {
               V:  "us-central1",
           },
           "name": {
               V:  "projects/pulumi-development/locations/us-central1/connectionProfiles/db_bigquery-replication-cp-FZz3PEgE6Hx3iUgi",
           },
           "project": {
               V:  "pulumi-development",
           },
       }
```

Notice there's no `bigquery_profile` parameter at al.

**Comment by VenelinMartinov** [comment]
Dumped some data from the Create method.

Looks like the [TF apply](https://github.com/pulumi/pulumi-terraform-bridge/blob/d744c5d79cde19b44ff7999f9a68918cca1bd596/pkg/tfbridge/provider.go#L1050) correctly returns `"bigquery_profile": []interface {}{ map[string]interface {}{}}`. Then [MakeTerraformResult](https://github.com/pulumi/pulumi-terraform-bridge/blob/d744c5d79cde19b44ff7999f9a68918cca1bd596/pkg/tfbridge/provider.go#L1074) returns `"bigqueryProfile": {}` which seems sensible and [MarshallProperties](https://github.com/pulumi/pulumi-terraform-bridge/blob/d744c5d79cde19b44ff7999f9a68918cca1bd596/pkg/tfbridge/provider.go#L1089) seems to drop it:

```golang
"bigqueryProfile": &structpb.Value{
     state:         impl.MessageState{},
     sizeCache:     0,
     unknownFields: nil,
     Kind:          &structpb.Value_NullValue{},
 }
```

So the issue is either in MarshallProperties or in the output of MakeTerraformResult.

**Comment by VenelinMartinov** [comment]
GCP CI is blocked on https://github.com/pulumi/pulumi-gcp/issues/1888, we also need to commit and release https://github.com/pulumi/pulumi-terraform-bridge/pull/1826 and get it propagated to the bridge. This is likely to take a few days.

**Comment by VenelinMartinov** [comment]
Released in https://github.com/pulumi/pulumi-gcp/releases/tag/v7.18.0

**Comment by breno-bicycle** [comment]
Thank you for working on this!!

**Comment by emilong** [comment]
I'm experiencing a very similar issue going from 6.62.0 to 7.24.0, so theoretically I should have the benefited from the fix. The only difference in the situation I can see from the OP is that I have not tried destroying and recreating the resource, so it is possible there are still migration issues going directly from 6 to 7? (or perhaps some other context I missed in doing the upgrade?)

**Comment by emilong** [comment]
Update: My colleague was able to get things working again by doing a `pulumi stack export > stack.json`, wherein the `bigqueryProfile` value was `null`, editing the value to be `{}`, then doing a `stack import`.

Not sure if this is because of this library specifically or something further down the stack, but I thought this might be helpful to share.

**Comment by VenelinMartinov** [comment]
Hey @emilong, sorry to hear you had troubles here and glad you figured a way around.

I'd really appreciate it if you could open a new issue with the details of the issues you had as well as a short program which reproduces it 🙏 Were you also using the `bigqueryProfile` on the `ConnectionProfile` resource?

**Comment by emilong** [comment]
Thanks for the response, @VenelinMartinov! I'm pretty new to pulumi (basically just debugging an existing usage I didn't set up), so getting a repro together might take me a while, but I'll see what I can do. If I'm able to do that, I'll open the issue then.
